src/controllers/discussionController.ts

```diff
@@ -0,0 +1,266 @@
+import { Request, Response, NextFunction } from "express";
+import Discussion, { IDiscussion } from "../models/Discussion";
+import { StatusCodes } from "http-status-codes";
+import {
+  BadRequestError,
+  NotFoundError,
+  UnauthenticatedError,
+} from "../errors";
+import Joi from "joi";
+import mongoose from "mongoose";
+import { IUser } from "../models/User";
+
+const discussionJoiSchema = Joi.object({
+  title: Joi.string().required(),
+  book: Joi.string().required(),
+  content: Joi.string().required(),
+  date: Joi.date().iso().required(),
+  participants: Joi.array().items(Joi.string()).optional(),
+  meetingLink: Joi.string().uri().required(),
+  createdBy: Joi.string().required(),
+});
+
+interface IGetDiscussionsQuery {
+  search?: string;
+  sort?: "latest" | "oldest";
+}
+
+interface IJoinDiscussionBody {
+  user: mongoose.Schema.Types.ObjectId;
+}
+
+const getAllDiscussions = async (
+  req: Request<{}, {}, {}, IGetDiscussionsQuery & { bookId?: string }>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { search, sort, bookId } = req.query;
+
+  let query: any = {};
+
+  if (bookId) {
+    query.book = bookId;
+  }
+
+  if (search) {
+    query.title = { $regex: new RegExp(search, "i") };
+  }
+
+  const sortOptions: any = {};
+  if (sort) {
+    if (sort === "latest") sortOptions.date = -1;
+    else if (sort === "oldest") sortOptions.date = 1;
+  } else {
+    sortOptions.date = -1;
+  }
+
+  try {
+    const discussions = await Discussion.find(query).sort(sortOptions);
+    res.status(StatusCodes.OK).json({ discussions, count: discussions.length });
+  } catch (error) {
+    return next(error);
+  }
+};
+
+const getDiscussion = async (
+  req: Request<{ discussionId: string }>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { discussionId } = req.params;
```

Comment:
you can use validation whenever any input is received from the user, in this case: discussionId

Reply:
done